File: job-2022-autumn/378.kth-smallest-element-in-a-sorted-matrix.js
/*
 * @lc app=leetcode id=378 lang=javascript
 *
 * [378] Kth Smallest Element in a Sorted Matrix
 */

// @lc code=start
/**
 * @param {number[][]} matrix
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function (matrix, k) {
    const m = matrix.length;
    const n = matrix[0].length;

    let min = matrix[0][0];
    let max = matrix[m - 1][n - 1];

    while (true) {
        const mid = min + Math.floor((max - min) / 2);

        const leqCount = countLessEqualThanInMatrix(matrix, mid);

        if (leqCount >= k) {
            if (countLessEqualThanInMatrix(matrix, mid - 1) < k) {
                return mid;
            } else {
                max = mid - 1;
            }
        } else if (leqCount < k) {
            min = mid + 1;
        }
    }
};

/**
 * @param {number[][]} matrix
 * @param {number} target
 * @returns {number}
 */
function countLessEqualThanInMatrix(matrix, target) {
    let leqCount = 0;
    for (let i = 0; i < matrix.length; i++) {
        leqCount += countLessEqualThan(matrix[i], target);
    }
    return leqCount;
}

/**
 * @param {number[]} array
 * @param {number} target
 * @returns {number}
 */
function countLessEqualThan(array, target) {
    /**
     * @param {number} left
     * @param {number} right
     * @returns {number}
     */
    const helper = (left, right) => {
        if (left === right) {
            return left + 1;
        }

        const mid = left + Math.floor((right - left) / 2);

        if (array[mid] > target) {
            return helper(left, mid - 1);
        } else {
            if (array[mid + 1] > target) {
                return mid + 1;
            } else {
                return helper(mid + 1, right);
            }
        }
    };

    if (array[0] > target) {
        return 0;
    }
    if (array[array.length - 1] <= target) {
        return array.length;
    }

    return helper(0, array.length - 1);
}

// @lc code=end

File: Daily/378.kth-smallest-element-in-a-sorted-matrix.cpp
/*
 * @lc app=leetcode id=378 lang=cpp
 *
 * [378] Kth Smallest Element in a Sorted Matrix
 */
#include <vector>

using std::vector;

// @lc code=start
class Solution {
 public:
  int kthSmallest(vector<vector<int>>& matrix, int k) {
    const int M = matrix.size();
    const int N = matrix[0].size();

    int left = matrix[0][0];
    int right = matrix[M - 1][N - 1];
    int result = -1;

    // find smallest element has at least k smaller elements
    // https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/solutions/1322101/c-java-python-maxheap-minheap-binary-search-picture-explain-clean-concise/
    while (left <= right) {
      const int mid = (right - left) / 2 + left;
      if (countLessEqual(matrix, mid) >= k) {
        if (countLessEqual(matrix, mid - 1) < k) {
          result = mid;
          break;
        } else {
          right = mid - 1;
        }
      } else {
        left = mid + 1;
      }
    }

    return result;
  }

  int countLessEqual(vector<vector<int>>& matrix, const int num) const {
    const int M = matrix.size();
    const int N = matrix[0].size();
    int lessThanNumElementCount = 0;
    for (int i = 0; i < M && matrix[i][0] <= num; i++) {
      if (matrix[i][N - 1] <= num) {
        lessThanNumElementCount += N;
      } else {
        for (int j = 0; j < N; j++) {
          if (matrix[i][j] <= num) {
            lessThanNumElementCount++;
          } else {
            break;
          }
        }
      }
    }
    return lessThanNumElementCount;
  }
};
// @lc code=end

int main() {
  Solution sol;
  vector<vector<int>> vec = {{1, 2}, {1, 3}};
  sol.kthSmallest(vec, 2);
}